### DISCUSSION TITLE
Can mongoid push validators to mongo?

### DISCUSSION THREAD

**Discussion by gfody** [mongodb/mongoid#5803] [Q&A]
warning: I am new to mongoid - this may be a dumb question

in our model classes we have code like
```rb
index({ :some_field => 1 }, { :unique => true })
...
validates :some_field, :presence => true
```

I'm told that the `index` statements cause indexes to be created in mongo but the `validates` statements are only enforced locally.

Is there a statement that will cause validators to be created in mongo?

**Comment by alexbevi** [comment]
Hey @gfody, Mongoid's [Validation](https://www.mongodb.com/docs/mongoid/master/reference/validation/) is effectively the same as Rail's [Active Record Validations](https://guides.rubyonrails.org/active_record_validations.html).

As this is a client-side feature there's nothing MongoDB can enforce server-side. 

If your question is about server-side [schema validation](https://www.mongodb.com/docs/manual/core/schema-validation/) this is better handled when a collection is created (or collection metadata is managed).

You could potentially do this using Mongoid, but that may not be the best place for that logic to live.

**Reply by gfody** [reply]
yes server-side schema validation is what I'm interested in, but mongoid is all we have for managing collection metadata - I am curious what potential ways exist to do this w/mongoid

**Reply by alexbevi** [reply]
You could do this via Mongoid, but it's a bit of a hack:

```ruby
require 'bundler/inline'

gemfile do  
  source 'https://rubygems.org'

  gem 'mongoid'  
end

Mongoid.configure do |config|
  config.clients.default = {
    uri: "mongodb://..."
  }  
end

class Foo
  include Mongoid::Document

  field :some_field, type: String
end

Foo.create!
# => #<Foo _id: 65f995e5a7ea754199ee523a, some_field: nil>

Foo.collection.database.command({ collMod: "foos",
   validator: {
      "$jsonSchema": {
         bsonType: "object",
         required: [ "some_field" ],
         properties: {
          some_field: {
              bsonType: "string",
              description: "some_field must be a string and is required"
          },       
        }      
      }
   }
})

Foo.create!
# => Mongo::Error::OperationFailure: [121]: Document failed validation -- {"failingDocumentId":{"$oid":"65f9960ca7ea754199ee523b"},"details":{"operatorName":"$jsonSchema","schemaRulesNotSatisfied":[{"operatorName":"required","specifiedAs":{"required":["some_field"]},"missingProperties":["some_field"]}]}} (on 10.0.0.6:27017, modern retry, attempt 1)
```
